pm_tb_data/fetch/nrt.py

```diff
@@ -0,0 +1,186 @@
+"""Assess access options for NRT AMSR2 LANCE data."""
+import copy
+import datetime as dt
+import os
+import re
+from pathlib import Path
+from typing import Literal, TypedDict, cast
+
+import earthaccess
+import requests
+from earthaccess.results import DataGranule
+from loguru import logger
+
+from pm_tb_data.fetch.errors import FetchRemoteDataError
+
+_URS_COOKIE = "urs_user_already_logged"
+_CHUNK_SIZE = 8 * 1024
+
+
+def _get_earthdata_creds():
+    if not os.environ.get("EARTHDATA_USERNAME"):
+        raise RuntimeError("Environment variable EARTHDATA_USERNAME must be defined.")
+    if not os.environ.get("EARTHDATA_PASSWORD"):
+        raise RuntimeError("Environment variable EARTHDATA_PASSWORD must be defined.")
+
+    return (
+        os.environ["EARTHDATA_USERNAME"],
+        os.environ["EARTHDATA_PASSWORD"],
+    )
+
+
+def _create_earthdata_authenticated_session(s=None, *, hosts: list[str], verify):
+    if not s:
+        s = requests.session()
+
+    for host in hosts:
+        resp = s.get(
+            host,
+            # We only want to inspect the redirect, not follow it yet:
+            allow_redirects=False,
+            # We don't want to accidentally fetch any data:
+            stream=True,
+            verify=verify,
+        )
+        # Copy the headers so they can be used case-insensitively after the
+        # response is closed.
+        headers = {k.lower(): v for k, v in resp.headers.items()}
+        resp.close()
+
+        redirected = resp.status_code == 302
+        redirected_to_urs = (
+            redirected and "urs.earthdata.nasa.gov" in headers["location"]
+        )
+
+        if not (redirected_to_urs):
+            print(f"Host {host} did not redirect to URS -- continuing without auth.")
+            return s
+
+        auth_resp = s.get(
+            headers["location"],
+            # Don't download data!
+            stream=True,
+            auth=_get_earthdata_creds(),
+        )
+        resp.close()
+        if not (auth_resp.ok and s.cookies.get(_URS_COOKIE) == "yes"):
+            msg = f"Authentication with Earthdata Login failed with:\n{auth_resp.text}"
+            raise RuntimeError(msg)
+
+        print(f"Authenticated for {host} with Earthdata Login.")
+
+    return s
+
+
+FileType = Literal["R04", "P04"]
+
+
+class GranuleInfo(TypedDict):
+    file_type: FileType
+    filename: str
+    data_url: str
+
+
+GranuleInfoByDate = dict[dt.date, GranuleInfo]
+
+
+def _get_granule_info_by_date(*, data_granules: list[DataGranule]) -> GranuleInfoByDate:
+    fn_pattern = re.compile(
+        r"AMSR_U2_L3_SeaIce12km_(?P<file_type>P04|R04)_(?P<file_date>\d{8}).he5"
```

Comment:
I might be being pedantic, but the file_type is probably "P" or "R" with a version of "04".  I don't think they upversion very frequently, but we might want to use just the letter to identify file_type

Reply:
Good observation. Do you think it's worth raising an error if the version is changed? A version change may indicate an incompatible data structure/organization.

 I suppose we could still download files of a different version, and then try to access them with our existing code. If that fails we'd have to update the access code to support the new version, but the fetch code would continue to work.

Reply:
I do think we should "be notified" if the version changes.  

But I think it should be a warning and not an error.  

I would expect that the API to the data would not change -- despite what a major version change usually means -- but the "science team" of this project will certainly want to evaluate the situation if the file version number changes.

Reply:
Good idea - I'll log this as a warning instead of an error. We can handle unexpected data formatting in the accessor code.

Reply:
1dd86e5